fuel-core/src/state/in_memory/transaction.rs

```diff
@@ -30,44 +27,47 @@ use std::{
         Mutex,
     },
 };
+use strum::EnumCount;
+
+// use hashmap to collapse changes (e.g. insert then remove the same key)
+type ChangesMap = Mutex<HashMap<Vec<u8>, WriteOperation>>;
 
 #[derive(Debug)]
 pub struct MemoryTransactionView {
     view_layer: MemoryStore,
-    // use hashmap to collapse changes (e.g. insert then remove the same key)
-    changes: Arc<Mutex<HashMap<Vec<u8>, WriteOperation>>>,
+    changes: Arc<[ChangesMap; Column::COUNT]>,
     data_source: DataSource,
 }
 
 impl MemoryTransactionView {
     pub fn new(source: DataSource) -> Self {
+        let changes: [_; Column::COUNT] = Default::default();
         Self {
             view_layer: MemoryStore::default(),
-            changes: Default::default(),
+            changes: Arc::new(changes),
             data_source: source,
         }
     }
 
     pub fn commit(&self) -> crate::state::Result<()> {
-        self.data_source.batch_write(
-            &mut self
-                .changes
-                .lock()
+        let mut iter = self.changes.iter().flat_map(|lock| {
+            lock.lock()
```

Comment:
hmm 🤔 

Shouldn't we lock all the columns first before draining anything? Otherwise there could still be writes happening to the unlocked columns during a commit.